File: Android V1.6/app/src/main/java/com/arcinteractive/spaces/ui/screens/events/EventsViewModel.kt
package com.arcinteractive.spaces.ui.screens.events

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.arcinteractive.spaces.data.model.EditableSpaceEvent
import com.arcinteractive.spaces.data.model.EventEditorMode
import com.arcinteractive.spaces.data.model.Space
import com.arcinteractive.spaces.data.model.SpaceEvent
import com.arcinteractive.spaces.data.spaces.SpaceService
import com.google.firebase.firestore.ListenerRegistration
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class EventsUiState(
    val events: List<SpaceEvent> = emptyList(),
    val isLoading: Boolean = false,
    val canCreateEvents: Boolean = false,
    val selectedEvent: SpaceEvent? = null,
    val canManageSelectedEvent: Boolean = false,
    val editorMode: EventEditorMode? = null,
    val isSaving: Boolean = false,
    val isDeleting: Boolean = false,
    val errorMessage: String? = null
)

class EventsViewModel(
    val space: Space,
    private val spaceService: SpaceService = SpaceService()
) : ViewModel() {
    private val _uiState = MutableStateFlow(EventsUiState(isLoading = true))
    val uiState: StateFlow<EventsUiState> = _uiState.asStateFlow()

    private var eventsListener: ListenerRegistration? = null

    fun startListening(context: Context) {
        if (eventsListener != null) return

        _uiState.update { it.copy(isLoading = true) }
        viewModelScope.launch {
            _uiState.update {
                it.copy(canCreateEvents = spaceService.canPerform(context, space, com.arcinteractive.spaces.data.model.SpacePermission.CreateEvents))
            }
        }
        eventsListener = spaceService.listenToEvents(
            context = context,
            space = space,
            listenerKey = "events.${space.id}"
        ) { result ->
            result.onSuccess { events ->
                val sortedEvents = events.sortedBy { it.startDate.time }
                _uiState.update { current ->
                    current.copy(
                        events = sortedEvents,
                        isLoading = false,
                        selectedEvent = current.selectedEvent?.let { selected ->
                            sortedEvents.firstOrNull { it.id == selected.id }
                        }
                    )
                }
            }.onFailure { error ->
                _uiState.update {
                    it.copy(
                        isLoading = false,
                        errorMessage = error.localizedMessage ?: "Unable to load events."
                    )
                }
            }
        }
    }

    fun openEvent(context: Context, event: SpaceEvent) {
        _uiState.update { it.copy(selectedEvent = event) }
        refreshSelectedEventPermissions(context, event)
    }

    fun closeEvent() {
        _uiState.update { it.copy(selectedEvent = null, canManageSelectedEvent = false) }
    }

    fun presentCreateEvent() {
        if (!_uiState.value.canCreateEvents) return
        _uiState.update { it.copy(editorMode = EventEditorMode.Create) }
    }

    fun presentEditEvent(event: SpaceEvent) {
        _uiState.update { it.copy(editorMode = EventEditorMode.Edit(event)) }
    }

    fun dismissEditor() {
        _uiState.update { it.copy(editorMode = null) }
    }

    fun saveEvent(context: Context, draft: EditableSpaceEvent, event: SpaceEvent?) {
        if (_uiState.value.isSaving) return

        viewModelScope.launch {
            _uiState.update { it.copy(isSaving = true) }
            runCatching {
                if (event == null) {
                    spaceService.createEvent(
                        context = context,
                        space = space,
                        title = draft.title,
                        description = draft.description,
                        location = draft.location,
                        startDate = draft.startDate,
                        endDate = draft.endDate,
                        allDay = draft.allDay
                    )
                } else {
                    spaceService.updateEvent(
                        context = context,
                        space = space,
                        event = event,
                        title = draft.title,
                        description = draft.description,
                        location = draft.location,
                        startDate = draft.startDate,
                        endDate = draft.endDate,
                        allDay = draft.allDay
                    )
                }
            }.onSuccess {
                _uiState.update { it.copy(isSaving = false, editorMode = null) }
            }.onFailure { error ->
                _uiState.update {
                    it.copy(
                        isSaving = false,
                        errorMessage = error.localizedMessage ?: "Unable to save event."
                    )
                }
            }
        }
    }

    fun deleteEvent(context: Context, event: SpaceEvent) {
        if (_uiState.value.isDeleting) return

        viewModelScope.launch {
            _uiState.update { it.copy(isDeleting = true) }
            runCatching {
                spaceService.deleteEvent(context, space, event)
            }.onSuccess {
                _uiState.update {
                    it.copy(
                        isDeleting = false,
                        selectedEvent = null,
                        canManageSelectedEvent = false
                    )
                }
            }.onFailure { error ->
                _uiState.update {
                    it.copy(
                        isDeleting = false,
                        errorMessage = error.localizedMessage ?: "Unable to delete event."
                    )
                }
            }
        }
    }

    fun clearError() {
        _uiState.update { it.copy(errorMessage = null) }
    }

    private fun refreshSelectedEventPermissions(context: Context, event: SpaceEvent) {
        viewModelScope.launch {
            val canManage = spaceService.canManageEvent(context, space, event)
            _uiState.update { current ->
                if (current.selectedEvent?.id == event.id) {
                    current.copy(canManageSelectedEvent = canManage)
                } else {
                    current
                }
            }
        }
    }

    override fun onCleared() {
        eventsListener?.remove()
        eventsListener = null
        super.onCleared()
    }
}

File: iOS/Spaces V1.5/ViewModels/EventsViewModel.swift
import Combine
import FirebaseFirestore
import Foundation

@MainActor
final class EventsViewModel: ObservableObject {
    @Published private(set) var events: [SpaceEvent] = []
    @Published private(set) var isLoading = false
    @Published private(set) var selectedEvent: SpaceEvent?
    @Published var editorMode: EventEditorMode?
    @Published var errorMessage: String?
    @Published var isSaving = false
    @Published var isDeleting = false
    @Published private(set) var canCreateEvents = false

    let space: Space

    private let spaceService: SpaceService
    private var eventsListener: ListenerRegistration?

    init(space: Space, spaceService: SpaceService? = nil) {
        self.space = space
        self.spaceService = spaceService ?? SpaceService()
        startListening()
    }

    deinit {
        eventsListener?.remove()
    }

    func startListening() {
        guard eventsListener == nil else { return }
        isLoading = true
        Task {
            canCreateEvents = await spaceService.canPerform(.createEvents, in: space)
        }
        eventsListener = spaceService.listenToEvents(in: space) { [weak self] result in
            guard let self else { return }
            switch result {
            case .success(let events):
                self.events = events.sorted { $0.sortDate < $1.sortDate }
                self.selectedEvent = self.selectedEvent.flatMap { selected in
                    self.events.first(where: { $0.id == selected.id })
                }
                self.isLoading = false
            case .failure(let error):
                self.errorMessage = error.localizedDescription
                self.isLoading = false
            }
        }
    }

    func presentCreateEvent() {
        guard canCreateEvents else { return }
        editorMode = .create
    }

    func presentEditEvent(_ event: SpaceEvent) {
        editorMode = .edit(event)
    }

    func dismissEditor() {
        editorMode = nil
    }

    func openEvent(_ event: SpaceEvent) {
        selectedEvent = event
    }

    func closeEvent() {
        selectedEvent = nil
    }

    func saveEvent(_ draft: EditableSpaceEvent, editing event: SpaceEvent?) async -> Bool {
        guard !isSaving else { return false }
        isSaving = true
        defer { isSaving = false }

        do {
            if let event {
                try await spaceService.updateEvent(
                    in: space,
                    event: event,
                    title: draft.title,
                    description: draft.description,
                    location: draft.location,
                    startDate: draft.startDate,
                    endDate: draft.endDate,
                    allDay: draft.allDay
                )
            } else {
                _ = try await spaceService.createEvent(
                    in: space,
                    title: draft.title,
                    description: draft.description,
                    location: draft.location,
                    startDate: draft.startDate,
                    endDate: draft.endDate,
                    allDay: draft.allDay
                )
            }
            editorMode = nil
            return true
        } catch {
            errorMessage = error.localizedDescription
            return false
        }
    }

    func deleteEvent(_ event: SpaceEvent) async {
        guard !isDeleting else { return }
        isDeleting = true
        defer { isDeleting = false }

        do {
            try await spaceService.deleteEvent(in: space, event: event)
            if selectedEvent?.id == event.id {
                selectedEvent = nil
            }
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func clearError() {
        errorMessage = nil
    }
}
